dask/dataframe/operation.py

```diff
@@ -0,0 +1,596 @@
+from __future__ import annotations
+
+import itertools
+import operator
+from collections.abc import Mapping
+from functools import singledispatch
+from typing import Any
+
+import numpy as np
+
+from dask.base import tokenize
+from dask.dataframe.utils import make_meta
+from dask.highlevelgraph import HighLevelGraph
+from dask.operation import CollectionOperation, MemoizingVisitor, operations, regenerate
+from dask.utils import apply, is_arraylike
+
+
+class DataFrameOperation(CollectionOperation):
+
+    _name: str
+    _meta: Any
+    _divisions: tuple
+    _columns: set | None = None
+    applied_filters: list | None = None
+    projected_columns: set | None = None
```

Comment:
Does it make sense for these properties to be represented by separate `Operation` objects, rather than being properties of every `DataFrameOperation`?

e.g.
```
class ProjectColumns:
    def __init__(self, op, *columns):
        self.op = op
        self._columns = frozenset(columns)
```

Now you one can also make the `DataFrameOperation` hashable (right now it isn't since some of its properties are not)

Then, column projection pushes the `ProjectColumn` operation as close to the leaves as is possible (hopefully one ends up with `ProjectColumn(SomeDataFrame, "column_name"))`, and generation of the lower-level graph from this can fuse these two things.